tests/apollo/test_skvbc_cron.py

```diff
@@ -0,0 +1,73 @@
+# Concord
+#
+# Copyright (c) 2021 VMware, Inc. All Rights Reserved.
+#
+# This product is licensed to you under the Apache 2.0 license (the "License").
+# You may not use this product except in compliance with the Apache 2.0 License.
+#
+# This product may include a number of subcomponents with separate copyright
+# notices and license terms. Your use of these subcomponents is subject to the
+# terms and conditions of the subcomponent's license, as noted in the LICENSE
+# file.
+
+import os.path
+import unittest
+
+import trio
+
+from util.bft import with_trio, with_bft_network, KEY_FILE_PREFIX
+
+expected_number_of_executes = 3
+
+def start_replica_cmd(builddir, replica_id):
+    """
+    Return a command that starts an skvbc replica when passed to
+    subprocess.Popen.
+
+    Note each arguments is an element in a list.
+    """
+    status_timer_milli = "500"
+    view_change_timeout_milli = "10000"
+    path = os.path.join(builddir, "tests", "simpleKVBC", "TesterReplica", "skvbc_replica")
+    return [path,
+            "-k", KEY_FILE_PREFIX,
+            "-i", str(replica_id),
+            "-s", status_timer_milli,
+            "-v", view_change_timeout_milli,
+            "-e", str(True),
+            "-r", str(expected_number_of_executes)
+            ]
+
+class SkvbcTestCron(unittest.TestCase):
+    __test__ = False  # so that PyTest ignores this test scenario
+
+    period_after_executes_seconds = 7
+    expected_component_id = "42"
+
+    @with_trio
+    @with_bft_network(start_replica_cmd, rotate_keys=True)
+    async def test_cron_action_called_correct_number_of_times(self, bft_network):
+        """
+        Test that Concord Cron actions are called the correct number of times.
+
+        Instruct TesterReplica to add a cron entry that is executed a specific
+        number of times. Verify that number after an additional period of time,
+        ensuring that the cron table doesn't execute the entry any more, even
+        though ticks generation is proceeding.
+        """
+        bft_network.start_all_replicas()
+
+        # Since the cron entry is called every second, wait for the number of expected executes + a period after them.
+        await trio.sleep(seconds=expected_number_of_executes + self.period_after_executes_seconds)
```

Comment:
This is non-deterministic. The [wait_for](https://github.com/vmware/concord-bft/blob/master/tests/apollo/util/bft.py#L1194-L1217) pattern should be used instead of a sleep, otherwise the metric checks below may fail.

Reply:
Oh, cool, I wasn't aware of that function! Will use it, thanks.

Reply:
Actually, `get_metric()` seems to be doing the same thing, just with hardcoded parameters. Still, I agree `wait_for()` is better as it takes the test case into consideration, allowing for custom timeout and interval. The `trio.sleep()` I've put seems to be redundant anyway.

Will redo the code with `wait_for()`

Reply:
`get_metric` doesn't do the same thing. I thought someone added that at first glance also. It only checks if no request has been received or the metric isn't yet published. In essence it helps with slow startup issues. `wait_for` allows you to wait specifically for a metric to reach an exact value.

Reply:
I see, but I was assuming I should be calling `wait_for()` with `retrieve_metric()` as a task. As an example:
https://github.com/vmware/concord-bft/blob/master/tests/apollo/util/bft.py#L1225

In that case, it seems to be doing the same thing, just with hardcoded time values.

I might be misunderstanding you or I might be missing something obvious, just haven't used Apollo a lot.

Reply:
Instead of using asserts below in the for loop, you pass a closure to `wait_for` that runs the loop and returns true only if `num_executes == expected_number_of_executes` and `component_id == self.expected_component_id` for all replicas.

Reply:
Actually, the task predicate relies on not being `None` for true, but the principle is the same. You loop checking that the conditions you want are true until they are or there is a timeout.

Reply:
Ah I see. That test looks buggy to me. A better example is this one: https://github.com/vmware/concord-bft/blob/master/tests/apollo/util/bft.py#L1124-L1133

Reply:
Thanks for the comment. I have removed the sleep and made the test independent of timing. I am waiting until the correct number of executes happens and I am using the last executed BFT sequence number to know the system has progressed.